docs/manpages/nmbd.8
.TH NMBD 8 17/1/1995 nmbd nmbd
.SH NAME
nmbd \- provide netbios nameserver support to clients
.SH SYNOPSIS
.B nmbd
[
.B -B
.I broadcast address
] [
.B -I
.I IP address
] [
.B -D
] [
.B -C comment string
] [
.B -G
.I group name
] [
.B -H
.I netbios hosts file
] [
.B -N
.I netmask
] [
.B -d
.I debuglevel
] [
.B -l
.I log basename
] [
.B -n
.I netbios name
] [
.B -p
.I port number
] [
.B -s
.I config file name
]

.SH DESCRIPTION
This program is part of the Samba suite.

.B nmbd
is a server that understands and can reply to netbios
name service requests, like those produced by LanManager
clients. It also controls browsing.

LanManager clients, when they start up, may wish to locate a LanManager server.
That is, they wish to know what IP number a specified host is using.

This program simply listens for such requests, and if its own name is specified
it will respond with the IP number of the host it is running on. "Its own name"
is by default the name of the host it is running on, but this can be overriden
with the
.B -n
option (see "OPTIONS" below). Using the

Nmbd can also be used as a WINS (Windows Internet Name Server)
server. It will do this automatically by default. What this basically
means is that it will respond to all name requests that it receives
that are not broadcasts, as long as it can resolve the name.
.SH OPTIONS
.B -B

.RS 3
On some systems, the server is unable to determine the broadcast address to
use for name registration requests. If your system has this difficulty, this 
parameter may be used to specify an appropriate broadcast address. The 
address should be given in standard "a.b.c.d" notation.

Only use this parameter if you are sure that the server cannot properly 
determine the proper broadcast address.

The default broadcast address is determined by the server at run time. If it
encounters difficulty doing so, it makes a guess based on the local IP
number.
.RE
.B -I

.RS 3
On some systems, the server is unable to determine the correct IP
address to use. This allows you to override the default choice.
.RE

.B -D

.RS 3
If specified, this parameter causes the server to operate as a daemon. That is,
it detaches itself and runs in the background, fielding requests on the 
appropriate port.

By default, the server will NOT operate as a daemon.
.RE

.B -C comment string

.RS 3
This allows you to set the "comment string" that is shown next to the
machine name in browse listings. 

A %v will be replaced with the Samba version number.

A %h will be replaced with the hostname.

It defaults to "Samba %v".
.RE

.B -G

.RS 3
This option allows you to specify a netbios group (also known as
lanmanager domain) that the server should be part of. You may include
several of these on the command line if you like. Alternatively you
can use the -H option to load a netbios hosts file containing domain names.

At startup, unless the -R switch has been used, the server will
attempt to register all group names in the hosts file and on the
command line (from the -G option).

The server will also respond to queries on this name.
.RE

.B -H

.RS 3
It may be useful in some situations to be able to specify a list of
netbios names for which the server should send a reply if
queried. This option allows that. The syntax is similar to the
standard /etc/hosts file format, but has some extensions.

The file contains three columns. Lines beginning with a # are ignored
as comments. The first column is an IP address, or a hostname. If it
is a hostname then it is interpreted as the IP address returned by
gethostbyname() when read. Any IP address of 0.0.0.0 will be
interpreted as the servers own IP address.

The second column is a netbios name. This is the name that the server
will respond to. It must be less than 20 characters long.

The third column is optional, and is intended for flags. Currently the
only flags supported are G, S and M. A G indicates that the name is a
group (also known as domain) name.

At startup all groups known to the server (either from this file or
from the -G option) are registered on the network (unless the -R
option has been selected).

A S or G means that the specified address is a broadcast address of a
network that you want people to be able to browse you from. Nmbd will
search for a master browser in that domain and will send host
announcements to that machine, informing it that the specifed somain
is available.

A M means that this name is the default netbios name for this
machine. This has the same affect as specifying the -n option to nmbd.

After startup the server waits for queries, and will answer queries to
any name known to it. This includes all names in the netbios hosts
file (if any), it's own name, and any names given with the -G option.

The primary intention of the -H option is to allow a mapping from
netbios names to internet domain names, and to allow the specification
of groups that the server should be part of.

.B Example:

        # This is a sample netbios hosts file

        # DO NOT USE THIS FILE AS-IS
        # YOU MAY INCONVENIENCE THE OWNERS OF THESE IPs
        # if you want to include a name with a space in it then 
        # use double quotes.

        # first put ourselves in the group LANGROUP
        0.0.0.0 LANGROUP G

        # next add a netbios alias for a faraway host
        arvidsjaur.anu.edu.au ARVIDSJAUR

        # finally put in an IP for a hard to find host
        130.45.3.213 FREDDY

        # now we want another subnet to be able to browse
        # us in the workgroup UNIXSERV
        192.0.2.255  UNIXSERV G

.RE

.B -M
.I workgroup name

.RS 3
If this parameter is given, the server will look for a master browser
for the specified workgroup name, report success or failure, then
exit. If successful, the IP address of the name located will be
reported. 

If you use the workgroup name "-" then nmbd will search for a master
browser for any workgroup by using the name __MSBROWSE__.

This option is meant to be used interactively on the command line, not
as a daemon or in inetd.

.RE
.B -N

.RS 3
On some systems, the server is unable to determine the netmask. If
your system has this difficulty, this parameter may be used to specify
an appropriate netmask. The mask should be given in standard
"a.b.c.d" notation.

Only use this parameter if you are sure that the server cannot properly 
determine the proper netmask.

The default netmask is determined by the server at run time. If it
encounters difficulty doing so, it makes a guess based on the local IP
number.
.RE

.B -d
.I debuglevel
.RS 3

debuglevel is an integer from 0 to 5.

The default value if this parameter is not specified is zero.

The higher this value, the more detail will be logged to the log files about
the activities of the server. At level 0, only critical errors and serious 
warnings will be logged. Level 1 is a reasonable level for day to day running
- it generates a small amount of information about operations carried out.

Levels above 1 will generate considerable amounts of log data, and should 
only be used when investigating a problem. Levels above 3 are designed for 
use only by developers and generate HUGE amounts of log data, most of which 
is extremely cryptic.
.RE

.B -l
.I log file

.RS 3
If specified,
.I logfile
specifies a base filename into which operational data from the running server
will be logged.

The default base name is specified at compile time.

The base name is used to generate actual log file names. For example, if the
name specified was "log", the following files would be used for log data:

.RS 3
log.nmb (containing debugging information)

log.nmb.in (containing inbound transaction data)

log.nmb.out (containing outbound transaction data)
.RE

The log files generated are never removed by the server.
.RE
.RE

.B -n
.I netbios name

.RS 3
This parameter tells the server what netbios name to respond with when 
queried. The same name is also registered on startup unless the -R 
parameter was specified.

The default netbios name used if this parameter is not specified is the 
name of the host on which the server is running.
.RE

.B -p
.I port number
.RS 3

port number is a positive integer value.

The default value if this parameter is not specified is 137.

This number is the port number that will be used when making connections to
the server from client software. The standard (well-known) port number for the
server is 137, hence the default. If you wish to run the server as an ordinary
user rather than as root, most systems will require you to use a port number
greater than 1024 - ask your system administrator for help if you are in this
situation.

Note that the name server uses UDP, not TCP!

This parameter is not normally specified except in the above situation.
.RE
.SH FILES

.B /etc/inetd.conf

.RS 3
If the server is to be run by the inetd meta-daemon, this file must contain
suitable startup information for the meta-daemon. See the section 
"INSTALLATION" below.
.RE

.B /etc/rc.d/rc.inet2

.RS 3
(or whatever initialisation script your system uses)

If running the server as a daemon at startup, this file will need to contain
an appropriate startup sequence for the server. See the section "Installation"
below.
.RE

.B /etc/services

.RS 3
If running the server via the meta-daemon inetd, this file must contain a
mapping of service name (eg., netbios-ns)  to service port (eg., 137) and
protocol type (eg., udp). See the section "INSTALLATION" below.
.RE
.RE

.SH ENVIRONMENT VARIABLES
Not applicable.

.SH INSTALLATION
The location of the server and its support files is a matter for individual
system administrators. The following are thus suggestions only.

It is recommended that the server software be installed under the /usr/local
hierarchy, in a directory readable by all, writeable only by root. The server
program itself should be executable by all, as users may wish to run the 
server themselves (in which case it will of course run with their privileges).
The server should NOT be setuid or setgid!

The server log files should be put in a directory readable and writable only
by root, as the log files may contain sensitive information.

The remaining notes will assume the following:

.RS 3
nmbd (the server program) installed in /usr/local/smb

log files stored in /var/adm/smblogs
.RE

The server may be run either as a daemon by users or at startup, or it may
be run from a meta-daemon such as inetd upon request. If run as a daemon, the
server will always be ready, so starting sessions will be faster. If run from 

NEWS
Changes in release 0.3d

 * add krb5-config

 * fix a bug in 3des gss-api mechanism, making it compatible with the
   specification and the MIT implementation

 * make telnetd only allow a specific list of environment variables to
   stop it from setting `sensitive' variables

 * try to use an existing libdes

 * lib/krb5, kdc: use correct usage type for ap-req messages.  This
   should improve compatability with MIT krb5 when using 3DES
   encryption types

 * kdc: fix memory allocation problem

 * update config.guess and config.sub

 * lib/roken: more stuff implemented

 * bug fixes and portability enhancements

Changes in release 0.3c

 * lib/krb5: memory caches now support the resolve operation

 * appl/login: set PATH to some sane default

 * kadmind: handle several realms

 * bug fixes (including memory leaks)

Changes in release 0.3b

 * kdc: prefer default-salted keys on v5 requests

 * kdc: lowercase hostnames in v4 mode

 * hprop: handle more types of MIT salts

 * lib/krb5: fix memory leak

 * bug fixes

Changes in release 0.3a:

 * implement arcfour-hmac-md5 to interoperate with W2K

 * modularise the handling of the master key, and allow for other
   encryption types. This makes it easier to import a database from
   some other source without having to re-encrypt all keys.

 * allow for better control over which encryption types are created

 * make kinit fallback to v4 if given a v4 KDC

 * make klist work better with v4 and v5, and add some more MIT
   compatibility options

 * make the kdc listen on the krb524 (4444) port for compatibility
   with MIT krb5 clients

 * implement more DCE/DFS support, enabled with --enable-dce, see
   lib/kdfs and appl/dceutils

 * make the sequence numbers work correctly

 * bug fixes

Changes in release 0.2t:

 * bug fixes

Changes in release 0.2s:

 * add OpenLDAP support in hdb

 * login will get v4 tickets when it receives forwarded tickets

 * xnlock supports both v5 and v4

 * repair source routing for telnet

 * fix building problems with krb4 (krb_mk_req)

 * bug fixes

Changes in release 0.2r:

 * fix realloc memory corruption bug in kdc

 * `add --key' and `cpw --key' in kadmin

 * klist supports listing v4 tickets

 * update config.guess and config.sub

 * make v4 -> v5 principal name conversion more robust

 * support for anonymous tickets

 * new man-pages

 * telnetd: do not negotiate KERBEROS5 authentication if there's no keytab.

 * use and set expiration and not password expiration when dumping
   to/from ka server databases / krb4 databases

 * make the code happier with 64-bit time_t

 * follow RFC2782 and by default do not look for non-underscore SRV names

Changes in release 0.2q:

 * bug fix in tcp-handling in kdc

 * bug fix in expand_hostname

Changes in release 0.2p:

 * bug fix in `kadmin load/merge'

 * bug fix in krb5_parse_address

Changes in release 0.2o:

 * gss_{import,export}_sec_context added to libgssapi

 * new option --addresses to kdc (for listening on an explicit set of
   addresses)

 * bug fixes in the krb4 and kaserver emulation part of the kdc

 * other bug fixes

Changes in release 0.2n:

 * more robust parsing of dump files in kadmin
 * changed default timestamp format for log messages to extended ISO
   8601 format (Y-M-DTH:M:S)
 * changed md4/md5/sha1 APIes to be de-facto `standard'
 * always make hostname into lower-case before creating principal
 * small bits of more MIT-compatability
 * bug fixes

Changes in release 0.2m:

 * handle glibc's getaddrinfo() that returns several ai_canonname

 * new endian test

 * man pages fixes

Changes in release 0.2l:

 * bug fixes

Changes in release 0.2k:

 * better IPv6 test

 * make struct sockaddr_storage in roken work better on alphas

 * some missing [hn]to[hn]s fixed.

 * allow users to change their own passwords with kadmin (with initial
   tickets)

 * fix stupid bug in parsing KDC specification

 * add `ktutil change' and `ktutil purge'

Changes in release 0.2j:

 * builds on Irix

 * ftpd works in passive mode

 * should build on cygwin

 * work around broken IPv6-code on OpenBSD 2.6, also add configure
   option --disable-ipv6

Changes in release 0.2i:

 * use getaddrinfo in the missing places.

 * fix SRV lookup for admin server

 * use get{addr,name}info everywhere.  and implement it in terms of
   getipnodeby{name,addr} (which uses gethostbyname{,2} and
   gethostbyaddr)

Changes in release 0.2h:

 * fix typo in kx (now compiles)

Changes in release 0.2g:

 * lots of bug fixes:
   * push works
   * repair appl/test programs
   * sockaddr_storage works on solaris (alignment issues)
   * works better with non-roken getaddrinfo
   * rsh works
   * some non standard C constructs removed

Changes in release 0.2f:

 * support SRV records for kpasswd
 * look for both _kerberos and krb5-realm when doing host -> realm mapping

Changes in release 0.2e:

 * changed copyright notices to remove `advertising'-clause.
 * get{addr,name}info added to roken and used in the other code
   (this makes things work much better with hosts with both v4 and v6
    addresses, among other things)
 * do pre-auth for both password and key-based get_in_tkt
 * support for having several databases
 * new command `del_enctype' in kadmin

source4/nmbd/nmbd.c
/*
   Unix SMB/CIFS implementation.
   NBT netbios routines and daemon - version 2
   Copyright (C) Andrew Tridgell 1994-1998
   Copyright (C) Jeremy Allison 1997-2002
   Copyright (C) Jelmer Vernooij 2002 (Conversion to popt)
   
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
   
*/

#include "includes.h"

int ClientNMB       = -1;
int ClientDGRAM     = -1;
int global_nmb_port = -1;

extern BOOL global_in_nmbd;

/* are we running as a daemon ? */
static BOOL is_daemon = False;

/* fork or run in foreground ? */
static BOOL Fork = True;

/* log to standard output ? */
static BOOL log_stdout = False;

/* have we found LanMan clients yet? */
BOOL found_lm_clients = False;

/* what server type are we currently */

time_t StartupTime = 0;

/**************************************************************************** **
 Handle a SIGTERM in band.
 **************************************************************************** */

static void terminate(void)
{
	DEBUG(0,("Got SIGTERM: going down...\n"));
  
	/* Write out wins.dat file if samba is a WINS server */
	wins_write_database(False);
  
	/* Remove all SELF registered names from WINS */
	release_wins_names();
  
	/* Announce all server entries as 0 time-to-live, 0 type. */
	announce_my_servers_removed();

	/* If there was an async dns child - kill it. */
	kill_async_dns_child();

	exit(0);
}

/**************************************************************************** **
 Handle a SHUTDOWN message from smbcontrol.
 **************************************************************************** */

static void nmbd_terminate(int msg_type, pid_t src, void *buf, size_t len)
{
	terminate();
}

/**************************************************************************** **
 Catch a SIGTERM signal.
 **************************************************************************** */

static SIG_ATOMIC_T got_sig_term;

static void sig_term(int sig)
{
	got_sig_term = 1;
	sys_select_signal();
}

/**************************************************************************** **
 Catch a SIGHUP signal.
 **************************************************************************** */

static SIG_ATOMIC_T reload_after_sighup;

static void sig_hup(int sig)
{
	reload_after_sighup = 1;
	sys_select_signal();
}

/*******************************************************************
 Print out all talloc memory info.
********************************************************************/

void return_all_talloc_info(int msg_type, pid_t src_pid, void *buf, size_t len)
{
	TALLOC_CTX *ctx = talloc_init("info context");
	char *info = NULL;

	if (!ctx)
		return;

	info = talloc_describe_all(ctx);
	if (info)
		DEBUG(10,(info));
	message_send_pid(src_pid, MSG_TALLOC_USAGE, info, info ? strlen(info) + 1 : 0, True);
	talloc_destroy(ctx);
}

#if DUMP_CORE
/**************************************************************************** **
 Prepare to dump a core file - carefully!
 **************************************************************************** */

static BOOL dump_core(void)
{
	char *p;
	pstring dname;
	pstrcpy( dname, lp_logfile() );
	if ((p=strrchr_m(dname,'/')))
		*p=0;
	pstrcat( dname, "/corefiles" );
	mkdir( dname, 0700 );
	sys_chown( dname, getuid(), getgid() );
	chmod( dname, 0700 );
	if ( chdir(dname) )
		return( False );
	umask( ~(0700) );

#ifdef HAVE_GETRLIMIT
#ifdef RLIMIT_CORE
	{
		struct rlimit rlp;
		getrlimit( RLIMIT_CORE, &rlp );
		rlp.rlim_cur = MAX( 4*1024*1024, rlp.rlim_cur );
		setrlimit( RLIMIT_CORE, &rlp );
		getrlimit( RLIMIT_CORE, &rlp );
		DEBUG( 3, ( "Core limits now %d %d\n", (int)rlp.rlim_cur, (int)rlp.rlim_max ) );
	}
#endif
#endif


	DEBUG(0,("Dumping core in %s\n",dname));
	abort();
	return( True );
}
#endif

/**************************************************************************** **
 Possibly continue after a fault.
 **************************************************************************** */

static void fault_continue(void)
{
#if DUMP_CORE
	dump_core();
#endif
}

/**************************************************************************** **
 Expire old names from the namelist and server list.
 **************************************************************************** */

static void expire_names_and_servers(time_t t)
{
	static time_t lastrun = 0;
  
	if ( !lastrun )
		lastrun = t;
	if ( t < (lastrun + 5) )
		return;
	lastrun = t;

	/*
	 * Expire any timed out names on all the broadcast
	 * subnets and those registered with the WINS server.
	 * (nmbd_namelistdb.c)
	 */

	expire_names(t);

	/*
	 * Go through all the broadcast subnets and for each
	 * workgroup known on that subnet remove any expired
	 * server names. If a workgroup has an empty serverlist
	 * and has itself timed out then remove the workgroup.
	 * (nmbd_workgroupdb.c)
	 */

	expire_workgroups_and_servers(t);
}

/************************************************************************** **
 Reload the list of network interfaces.
 ************************************************************************** */

static BOOL reload_interfaces(time_t t)
{
	static time_t lastt;
	int n;
	struct subnet_record *subrec;
	extern BOOL rescan_listen_set;
	extern struct in_addr loopback_ip;

	if (t && ((t - lastt) < NMBD_INTERFACES_RELOAD)) return False;
	lastt = t;

	if (!interfaces_changed()) return False;

	/* the list of probed interfaces has changed, we may need to add/remove
	   some subnets */
	load_interfaces();

	/* find any interfaces that need adding */
	for (n=iface_count() - 1; n >= 0; n--) {
		struct interface *iface = get_interface(n);

		/*
		 * We don't want to add a loopback interface, in case
		 * someone has added 127.0.0.1 for smbd, nmbd needs to
		 * ignore it here. JRA.
		 */

		if (ip_equal(iface->ip, loopback_ip)) {
			DEBUG(2,("reload_interfaces: Ignoring loopback interface %s\n", inet_ntoa(iface->ip)));
			continue;
		}

		for (subrec=subnetlist; subrec; subrec=subrec->next) {
			if (ip_equal(iface->ip, subrec->myip) &&
			    ip_equal(iface->nmask, subrec->mask_ip)) break;
		}

		if (!subrec) {
			/* it wasn't found! add it */
			DEBUG(2,("Found new interface %s\n", 
				 inet_ntoa(iface->ip)));
			subrec = make_normal_subnet(iface);
			if (subrec) register_my_workgroup_one_subnet(subrec);
		}
	}

	/* find any interfaces that need deleting */
	for (subrec=subnetlist; subrec; subrec=subrec->next) {
		for (n=iface_count() - 1; n >= 0; n--) {
			struct interface *iface = get_interface(n);
			if (ip_equal(iface->ip, subrec->myip) &&
			    ip_equal(iface->nmask, subrec->mask_ip)) break;
		}
		if (n == -1) {
			/* oops, an interface has disapeared. This is
			 tricky, we don't dare actually free the
			 interface as it could be being used, so
			 instead we just wear the memory leak and
			 remove it from the list of interfaces without
			 freeing it */
			DEBUG(2,("Deleting dead interface %s\n", 
				 inet_ntoa(subrec->myip)));
			close_subnet(subrec);
		}
	}
	
	rescan_listen_set = True;

	/* We need to shutdown if there are no subnets... */
	if (FIRST_SUBNET == NULL) {
		DEBUG(0,("reload_interfaces: No subnets to listen to. Shutting down...\n"));
		return True;
	}
	return False;
}

/**************************************************************************** **
 Reload the services file.
 **************************************************************************** */

static BOOL reload_nmbd_services(BOOL test)
{
	BOOL ret;

	set_remote_machine_name("nmbd");

	if ( lp_loaded() ) {
		pstring fname;
		pstrcpy( fname,lp_configfile());
		if (file_exist(fname,NULL) && !strcsequal(fname,dyn_CONFIGFILE)) {
			pstrcpy(dyn_CONFIGFILE,fname);
			test = False;
		}
	}

	if ( test && !lp_file_list_changed() )
		return(True);

	ret = lp_load( dyn_CONFIGFILE, True , False, False);

	/* perhaps the config filename is now set */
	if ( !test ) {
		DEBUG( 3, ( "services not loaded\n" ) );
		reload_nmbd_services( True );
	}

	return(ret);
}

/**************************************************************************** **
 The main select loop.
 **************************************************************************** */

static void process(void)
{
	BOOL run_election;

	while( True ) {
		time_t t = time(NULL);

		/* Check for internal messages */

		message_dispatch();

		/*
		 * Check all broadcast subnets to see if
		 * we need to run an election on any of them.
		 * (nmbd_elections.c)
		 */

		run_election = check_elections();

		/*
		 * Read incoming UDP packets.
		 * (nmbd_packets.c)
		 */

		if(listen_for_packets(run_election))
			return;

		/*
		 * Handle termination inband.
		 */

		if (got_sig_term) {
			got_sig_term = 0;
			terminate();
		}

		/*
		 * Process all incoming packets
		 * read above. This calls the success and
		 * failure functions registered when response
		 * packets arrrive, and also deals with request
		 * packets from other sources.
		 * (nmbd_packets.c)
		 */

		run_packet_queue();

		/*
		 * Run any elections - initiate becoming
		 * a local master browser if we have won.
		 * (nmbd_elections.c)
		 */

		run_elections(t);

		/*
		 * Send out any broadcast announcements
		 * of our server names. This also announces
		 * the workgroup name if we are a local
		 * master browser.
		 * (nmbd_sendannounce.c)
		 */

		announce_my_server_names(t);

		/*
		 * Send out any LanMan broadcast announcements
		 * of our server names.
		 * (nmbd_sendannounce.c)
		 */

		announce_my_lm_server_names(t);

		/*
		 * If we are a local master browser, periodically
		 * announce ourselves to the domain master browser.
		 * This also deals with syncronising the domain master
		 * browser server lists with ourselves as a local
		 * master browser.
		 * (nmbd_sendannounce.c)
		 */

		announce_myself_to_domain_master_browser(t);

		/*
		 * Fullfill any remote announce requests.
		 * (nmbd_sendannounce.c)
		 */

		announce_remote(t);

		/*
		 * Fullfill any remote browse sync announce requests.
		 * (nmbd_sendannounce.c)
		 */

		browse_sync_remote(t);

		/*
		 * Scan the broadcast subnets, and WINS client
		 * namelists and refresh any that need refreshing.
		 * (nmbd_mynames.c)
		 */

		refresh_my_names(t);

		/*
		 * Scan the subnet namelists and server lists and
		 * expire thos that have timed out.
		 * (nmbd.c)
		 */

		expire_names_and_servers(t);

		/*
		 * Write out a snapshot of our current browse list into
		 * the browse.dat file. This is used by smbd to service
		 * incoming NetServerEnum calls - used to synchronise
		 * browse lists over subnets.
		 * (nmbd_serverlistdb.c)
		 */

		write_browse_list(t, False);

		/*
		 * If we are a domain master browser, we have a list of
		 * local master browsers we should synchronise browse
		 * lists with (these are added by an incoming local
		 * master browser announcement packet). Expire any of
		 * these that are no longer current, and pull the server
		 * lists from each of these known local master browsers.
		 * (nmbd_browsesync.c)
		 */

		dmb_expire_and_sync_browser_lists(t);

		/*
		 * Check that there is a local master browser for our
		 * workgroup for all our broadcast subnets. If one
		 * is not found, start an election (which we ourselves
		 * may or may not participate in, depending on the
		 * setting of the 'local master' parameter.
		 * (nmbd_elections.c)
		 */

		check_master_browser_exists(t);

		/*
		 * If we are configured as a logon server, attempt to
		 * register the special NetBIOS names to become such
		 * (WORKGROUP<1c> name) on all broadcast subnets and
		 * with the WINS server (if used). If we are configured
		 * to become a domain master browser, attempt to register
		 * the special NetBIOS name (WORKGROUP<1b> name) to
		 * become such.
		 * (nmbd_become_dmb.c)
		 */

		add_domain_names(t);

		/*
		 * If we are a WINS server, do any timer dependent
		 * processing required.
		 * (nmbd_winsserver.c)
		 */

		initiate_wins_processing(t);

		/*
		 * If we are a domain master browser, attempt to contact the
		 * WINS server to get a list of all known WORKGROUPS/DOMAINS.
		 * This will only work to a Samba WINS server.
		 * (nmbd_browsesync.c)
		 */

		if (lp_enhanced_browsing())
			collect_all_workgroup_names_from_wins_server(t);

		/*
		 * Go through the response record queue and time out or re-transmit
		 * and expired entries.
		 * (nmbd_packets.c)
		 */

		retransmit_or_expire_response_records(t);

		/*
		 * check to see if any remote browse sync child processes have completed
		 */

		sync_check_completion();

		/*
		 * regularly sync with any other DMBs we know about 
		 */

		if (lp_enhanced_browsing())
			sync_all_dmbs(t);

		/*
		 * clear the unexpected packet queue 
		 */

		clear_unexpected(t);

		/*
		 * Reload the services file if we got a sighup.
		 */

		if(reload_after_sighup) {
			DEBUG( 0, ( "Got SIGHUP dumping debug info.\n" ) );
			write_browse_list( 0, True );
			dump_all_namelists();
			reload_nmbd_services( True );
			reopen_logs();
			if(reload_interfaces(0))
				return;
			reload_after_sighup = 0;
		}

		/* check for new network interfaces */

		if(reload_interfaces(t))
			return;

		/* free up temp memory */
			lp_talloc_free();
	}
}

/**************************************************************************** **
 Open the socket communication.
 **************************************************************************** */

static BOOL open_sockets(BOOL isdaemon, int port)
{
	/*
	 * The sockets opened here will be used to receive broadcast
	 * packets *only*. Interface specific sockets are opened in
	 * make_subnet() in namedbsubnet.c. Thus we bind to the
	 * address "0.0.0.0". The parameter 'socket address' is
	 * now deprecated.
	 */

	if ( isdaemon )
		ClientNMB = open_socket_in(SOCK_DGRAM, port,0,0,True);
	else
		ClientNMB = 0;
  
	ClientDGRAM = open_socket_in(SOCK_DGRAM,DGRAM_PORT,3,0,True);

	if ( ClientNMB == -1 )
		return( False );

	/* we are never interested in SIGPIPE */
	BlockSignals(True,SIGPIPE);

	set_socket_options( ClientNMB,   "SO_BROADCAST" );
	set_socket_options( ClientDGRAM, "SO_BROADCAST" );

	DEBUG( 3, ( "open_sockets: Broadcast sockets opened.\n" ) );
	return( True );
}

/**************************************************************************** **
 main program
 **************************************************************************** */
 int main(int argc, const char *argv[])
{
	static BOOL opt_interactive = False;
	poptContext pc;
	struct poptOption long_options[] = {
	POPT_AUTOHELP
	{"daemon", 'D', POPT_ARG_VAL, &is_daemon, True, "Become a daemon(default)" },
	{"interactive", 'i', POPT_ARG_VAL, &opt_interactive, True, "Run interactive (not a daemon)" },
	{"foreground", 'F', POPT_ARG_VAL, &Fork, False, "Run daemon in foreground (for daemontools & etc)" },
	{"log-stdout", 'S', POPT_ARG_VAL, &log_stdout, True, "Log to stdout" },
	{"hosts", 'H', POPT_ARG_STRING, dyn_LMHOSTSFILE, 'H', "Load a netbios hosts file"},
	{"port", 'p', POPT_ARG_INT, &global_nmb_port, NMB_PORT, "Listen on the specified port" },
	{NULL, 0, POPT_ARG_INCLUDE_TABLE, popt_common_debug },
	{NULL, 0, POPT_ARG_INCLUDE_TABLE, popt_common_configfile },
	{NULL, 0, POPT_ARG_INCLUDE_TABLE, popt_common_socket_options },
	{NULL, 0, POPT_ARG_INCLUDE_TABLE, popt_common_version },
	{NULL, 0, POPT_ARG_INCLUDE_TABLE, popt_common_netbios_name },
	{NULL, 0, POPT_ARG_INCLUDE_TABLE, popt_common_log_base },
	{ NULL }
	};
	int opt;
	pstring logfile;

  global_nmb_port = NMB_PORT;
  global_in_nmbd = True;

  StartupTime = time(NULL);

  sys_srandom(time(NULL) ^ sys_getpid());

  slprintf(logfile, sizeof(logfile)-1, "%s/log.nmbd", dyn_LOGFILEBASE);
  lp_set_logfile(logfile);

  fault_setup((void (*)(void *))fault_continue );

  /* POSIX demands that signals are inherited. If the invoking process has
   * these signals masked, we will have problems, as we won't recieve them. */
  BlockSignals(False, SIGHUP);
  BlockSignals(False, SIGUSR1);
  BlockSignals(False, SIGTERM);

  CatchSignal( SIGHUP,  SIGNAL_CAST sig_hup );
  CatchSignal( SIGTERM, SIGNAL_CAST sig_term );

#if defined(SIGFPE)
  /* we are never interested in SIGFPE */
  BlockSignals(True,SIGFPE);
#endif

  /* We no longer use USR2... */
#if defined(SIGUSR2)
  BlockSignals(True, SIGUSR2);
#endif
  pc = poptGetContext("nmbd", argc, argv, long_options, 0);
  
  while((opt = poptGetNextOpt(pc)) != -1)
    { }

  poptFreeContext(pc);

  if ( opt_interactive ) {
    Fork = False;
    log_stdout = True;
  }

  if ( log_stdout && Fork ) {
    DEBUG(0,("ERROR: Can't log to stdout (-S) unless daemon is in foreground (-F) or interactive (-i)\n"));
    exit(1);
  }

  setup_logging( argv[0], log_stdout?DEBUG_STDOUT : DEBUG_FILE );

  reopen_logs();

  DEBUG( 0, ( "Netbios nameserver version %s started.\n", VERSION ) );
  DEBUGADD( 0, ( "Copyright Andrew Tridgell and the Samba Team 1994-2002\n" ) );

  if ( !reload_nmbd_services(False) )
    return(-1);

  if(!init_names())
    return -1;

  reload_nmbd_services( True );

  if (strequal(lp_workgroup(),"*"))
  {
    DEBUG(0,("ERROR: a workgroup name of * is no longer supported\n"));
    exit(1);
  }

  set_samba_nb_type();

  if (!is_daemon && !is_a_socket(0))
  {
    DEBUG(0,("standard input is not a socket, assuming -D option\n"));
    is_daemon = True;
  }
  
  if (is_daemon && !opt_interactive)
  {
    DEBUG( 2, ( "Becoming a daemon.\n" ) );
    become_daemon(Fork);
  }

#if HAVE_SETPGID
  /*
   * If we're interactive we want to set our own process group for 
   * signal management.
   */
  if (opt_interactive)
    setpgid( (pid_t)0, (pid_t)0 );
#endif

#ifndef SYNC_DNS
  /* Setup the async dns. We do it here so it doesn't have all the other
     stuff initialised and thus chewing memory and sockets */
  if(lp_we_are_a_wins_server() && lp_dns_proxy()) {
	  start_async_dns();
  }
#endif

  if (!directory_exist(lp_lockdir(), NULL)) {
	  mkdir(lp_lockdir(), 0755);
  }

  pidfile_create("nmbd");
  message_init();
  message_register(MSG_FORCE_ELECTION, nmbd_message_election);
  message_register(MSG_WINS_NEW_ENTRY, nmbd_wins_new_entry);
  message_register(MSG_SHUTDOWN, nmbd_terminate);
  message_register(MSG_REQ_TALLOC_USAGE, return_all_talloc_info);

  DEBUG( 3, ( "Opening sockets %d\n", global_nmb_port ) );

  if ( !open_sockets( is_daemon, global_nmb_port ) ) {
    kill_async_dns_child();
    return 1;
  }

  /* Determine all the IP addresses we have. */
  load_interfaces();

  /* Create an nmbd subnet record for each of the above. */
  if( False == create_subnets() )
  {
    DEBUG(0,("ERROR: Failed when creating subnet lists. Exiting.\n"));
    kill_async_dns_child();
    exit(1);
  }

  /* Load in any static local names. */ 
  load_lmhosts_file(dyn_LMHOSTSFILE);
  DEBUG(3,("Loaded hosts file %s\n", dyn_LMHOSTSFILE));

  /* If we are acting as a WINS server, initialise data structures. */
  if( !initialise_wins() )
  {
    DEBUG( 0, ( "nmbd: Failed when initialising WINS server.\n" ) );
    kill_async_dns_child();
    exit(1);
  }

  /* 
   * Register nmbd primary workgroup and nmbd names on all
   * the broadcast subnets, and on the WINS server (if specified).
   * Also initiate the startup of our primary workgroup (start
   * elections if we are setup as being able to be a local
   * master browser.
   */

  if( False == register_my_workgroup_and_names() )
  {
    DEBUG(0,("ERROR: Failed when creating my my workgroup. Exiting.\n"));
    kill_async_dns_child();
    exit(1);
  }

  /* We can only take signals in the select. */
  BlockSignals( True, SIGTERM );

  process();

  if (dbf)
    x_fclose(dbf);
  kill_async_dns_child();
  return(0);
}
